platform/entanglement/src/services/LinkService.js

```diff
@@ -45,36 +45,24 @@ define(
             if (parentCandidate.getId() === object.getId()) {
                 return false;
             }
-            if (parentCandidate.getModel().composition.indexOf(object.getId()) !== -1) {
+            if ((parentCandidate.getModel().composition || []).indexOf(object.getId()) !== -1) {
                 return false;
             }
             return this.policyService.allow(
                 "composition",
                 parentCandidate.getCapability('type'),
                 object.getCapability('type')
-            );
+            ) && parentCandidate.hasCapability('composition');
         };
 
         LinkService.prototype.perform = function (object, parentObject) {
-            function findChild(children) {
-                var i;
-                for (i = 0; i < children.length; i += 1) {
-                    if (children[i].getId() === object.getId()) {
-                        return children[i];
-                    }
-                }
-            }
+            // Note that this was checked-for explicitly during validate step
```

Comment:
I think this is useful because it may not be immediately obvious that `validate` may/must be invoked before `perform`; will toy with phrasing a little to clarify that. (In general, it's risky to expose methods that need to be called in some particular sequence for correctness, so at bare minimum I like to have some kind of breadcrumb in case that goes wrong.)